### PR TITLE
Typescript port

### PR DESCRIPTION
**Pull request by cspotcode**
Per discussion in #132, I took a stab at porting to TypeScript.  Still WIP but the bulk of the renaming and reconfiguring is done.

@jedmao you might have ideas about this.

A more readable diff, which excludes all the renaming of .js to .ts, is here:
https://github.com/plopjs/node-plop/compare/eafd508cd975216c3535f973e3ed5e487f91cc03...cspotcode:typescript-port?expand=1

### Still to be done:
Remove all the redundant typings from `src/types.ts` (formerly index.d.ts)
Copy the relevant JSDoc from `types.ts` to the codebase proper
Annotate public API signatures with the same types as `types.ts`
Fix failing tests.
(optional) re-enable the disabled linting rules mentioned in `TYPESCRIPT_NOTES.md`

### CODE REVIEW

**Comment by jedmao** [comment]
Anything left on this?

**Comment by cspotcode** [comment]
The checklist in the PR description still needs to be done.  Looks like the
tests were fixed, though.

**Comment by cspotcode** [comment]
Feel free to run with it.  I don't have time right now.

**Comment by crutchcorn** [comment]
@amwmedia do you know if we, as maintainers, have a general consesus of it we really _want_ to migrate to TS?

I personally have no problem keeping it in JS and maintaining typings, but I'll defer to your judgement

**Comment by jacksteamdev** [comment]
@crutchcorn This is a good question. It's a huge job, and this PR is getting a bit stale.

At the very least, we should try to think of a way to verify that the typings work as intended. I've had some trouble working with actions. Maybe in the tests we run an example `plopfile.ts` file through `tsc` and verify that there are no errors?

**Comment by cspotcode** [comment]
I merged the latest master.  I also bumped node from 8.9 to 8.10 to meet the requirements of typescript-eslint.  We could still using 8.9 for unit tests if needed.

I've also attached a diff showing the code changes sans everything else.
It's about 35, 40 lines of source changed.  The rest of the changes are package-lock, files renamed .js -> .ts without any changes, or boilerplate like eslintrc.

[diff.txt](https://github.com/plopjs/node-plop/files/4258675/diff.txt)

**Comment by crutchcorn** [comment]
@cspotcode I've had experience setting up `dtslint` and can do so for this project. That said, is this merge ready-to-go other than the conflicting files that've just been added?

**Comment by cspotcode** [comment]
There's a checklist in the PRs description where I listed a few things that need to be done before this will spit out meaningful declarations.

I'm pretty sure you could merge this now, and then iterate over time to bring the emitted declarations up to parity with the manually-written declarations.  In the meantime, keep the manually written declarations for plop consumers to use.

That way the initial hurdle of renaming files would be done.  And you could take your time annotating the public API surface and making the JSDoc look nice.

File: src/generator-runner.ts [hunk 1]
```diff
@@ -152,3 +152,14 @@ export default function (plopfileApi, flags) {
...
 }
+
+interface ActionResult {
+	type: string;
+	path: string;
+}
+
+interface Failure {
+	type: string;
+	path: string;
```
Comment (jedmao) [inline src/generator-runner.ts]:
```suggestion
interface Failure extends ActionResult {
```

File: plopfile.ts [hunk 2]
```diff
@@ -1,4 +1,5 @@
-module.exports = function (plop) {
+export = factory;
+function factory(plop) {
```
Comment (jedmao) [inline plopfile.ts]:
```suggestion
export = function factory(plop) {
```

File: tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,71 @@
+{
+  "include": [
+    "src"
+  ],
+  "compilerOptions": {
+    /* Basic Options */
+    // "incremental": true,                   /* Enable incremental compilation */
+    "target": "es6",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019' or 'ESNEXT'. */
```
Comment (jedmao) [inline tsconfig.json]: Node 10 has [full ES2017 support](https://node.green/#ES2017). We should also bump the `engines` version in package.json to `node >= 10`, as 8 is out of LTS.

```suggestion
    "target": "ES2017",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019' or 'ESNEXT'. */
```

**Comment by cspotcode** [comment]
I enabled "allow edits from maintainers" so you can push to my branch, and if any non-maintainers want to run with this, feel free to move it to a new PR.

**Comment by crutchcorn** [comment]
@jacksteamdev being a larger contributor and having worked with the typings, I'll defer to your judgement. I'm personally leaning a bit more towards the "use `dtslint`" option simply because it seems to be a lessened amount of work and our codebase seems to work sufficiently without TypeScript, but I'm curious with what your thoughts are. 

Regardless of what we choose, I'll likely be working on solving the root problem of inconsistent typings sometime this weekend

**Comment by cspotcode** [comment]
Another option worth considering: they recently added the ability to emit
declarations from JS source, which can be paired with `checkJs` to ensure
declarations remain valid and in-sync with the codebase.  I don't like
working in those kinds of codebases, because forcing all type annotations
into JS comments is annoying and unnecessarily verbose, and I'd rather use
the native syntax.  But some teams prefer it.

**Comment by crutchcorn** [comment]
Having used `checkJs` to generate typings for a Google Maps package, I can comfortably say that it doesn't tend to generate the same level of type strictness, especially surrounding callback types. Not all of the JSDoc decorations are supported and in general I find it a lackluster DX. A fantastic feature when typings are not already present, but I'd rather manually enforce and write typings for `plop`  since we have a baseline

**Comment by jacksteamdev** [comment]
@crutchcorn I agree that `dtslint` is a solid choice for now. I use TS for all my new projects, but like you said, we have a solid base of working code here.

**Comment by jacksteamdev** [comment]
@crutchcorn @cspotcode That being said, I can see so much value in using TS, but it seems that we don't have enough people with enough time to get this done right now.

Should we open an issue to track community support for a TS conversion?

When we get several people who can see it through, we could come up with a plan to gradually do a full conversion in a way that the project would be ok if the conversion stalls for some time.

**Comment by cspotcode** [comment]
Just to clarify, if we merge this PR now, but keep the existing index.d.ts,
I'm pretty sure things continue to "just work." That's the best way to a
gradual conversion, because all the renaming is done and the tooling is in
place. (eslint, build step, ts-node in tests)

**Comment by jacksteamdev** [comment]
My main concern with a TS port is how it will affect contributions. Will having everything in TS discourage some from contributing? Not everyone is comfortable with TS. That's the kind of community support I meant.

**Comment by jedmao** [comment]
That argument might have stood ground a couple years ago, but as long as the project is setup with appropriate configuration and tooling in place, contributors will feel like they are writing JavaScript 95% of the time. This library doesn’t require crazy generics or anything. Plus, the community has warmed up to TypeScript over the years. They want it.

**Comment by amwmedia** [comment]
Sorry, I'm a little late to this discussion. I'm certainly open to the idea of a TS conversion, but I don't have enough TS experience to really contribute much. I'd be leaning on the team's expertise and time to get this done. It's a pretty big change, so if you guys feel that there's enough value and you're willing to put in the time to make it happen, I'm willing to treat it as a learning experience for myself. 😂

**Comment by cspotcode** [comment]
I'm happy to help with adding the necessary annotations after this gets merged.  I don't want to spend time on codebases that aren't TS, because I feel working in un-typed code takes longer than it needs to, so it's inefficient.  *Until* this gets merged, my focus is solely on making this PR merge-able, nothing more.  *After* it's merged I'll absolutely help migrate the necessary JSDoc.

I've once again merged the latest `master`, re-bundled the manually-written declarations, and added automated testing for both the manually written and emitted declarations.  This is the boilerplate we need to bring the emitted declarations to parity and eventually publish them.  In the meantime, the manually-written declarations will be bundled.

**Comment by crutchcorn** [comment]
Closing in favor of #166

**Pull request closed without merging**